tyrian/js/src/main/scala/tyrian/runtime/TyrianRuntime.scala

```diff
@@ -1,134 +1,97 @@
 package tyrian.runtime
 
+import cats.data.OptionT
 import cats.effect.kernel.Async
 import cats.effect.kernel.Ref
+import cats.effect.kernel.Resource
+import cats.effect.std.AtomicCell
 import cats.effect.std.Dispatcher
+import cats.effect.syntax.all.*
 import cats.syntax.all.*
 import fs2.Stream
 import fs2.concurrent.Channel
 import org.scalajs.dom
 import org.scalajs.dom.Element
 import snabbdom.VNode
-import tyrian.Attr
-import tyrian.Attribute
+import snabbdom.toVNode
 import tyrian.Cmd
-import tyrian.Event
 import tyrian.Html
 import tyrian.Sub
 
-final class TyrianRuntime[F[_]: Async, Model, Msg](
-    initCmd: Cmd[F, Msg],
-    update: Model => Msg => (Model, Cmd[F, Msg]),
-    view: Model => Html[Msg],
-    subscriptions: Model => Sub[F, Msg],
-    model: Ref[F, ModelHolder[Model]],
-    vnode: Ref[F, Element | VNode],
-    channel: => Channel[F, F[Unit]],
-    dispatcher: => Dispatcher[F]
-):
-
-  // The currently live subs.
-  @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
-  private var currentSubscriptions: List[(String, F[Unit])] = Nil
-  // This is a queue of new subs waiting to be run for the first time.
-  // In the event that two events happen at once, you can't assume that
-  // you would have run all the subs between events.
-  @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
-  private var aboutToRunSubscriptions: Set[String] = Set.empty
-
-  // Update
-
-  private def onMsg(msg: Msg): Unit =
-    val res: F[Unit] =
-      for {
-        currentModel <- model.get
-        (updatedState, cmd) = update(currentModel.model)(msg)
-        _ <- completeUpdate(cmd, updatedState)
-      } yield ()
-
-    dispatcher.unsafeRunAndForget(res)
-
-  private def completeUpdate(cmd: Cmd[F, Msg], updatedState: Model): F[Unit] =
-    for {
-      _           <- model.set(ModelHolder(updatedState, true))
-      sideEffects <- gatherSideEffects(cmd, subscriptions(updatedState))
-      _           <- Stream.emits(sideEffects).foreach(msg => channel.send(msg).void).compile.drain
-    } yield ()
+object TyrianRuntime:
+
+  def apply[F[_], Model, Msg](
+      node: Element,
+      initModel: Model,
+      initCmd: Cmd[F, Msg],
+      update: Model => Msg => (Model, Cmd[F, Msg]),
+      view: Model => Html[Msg],
+      subscriptions: Model => Sub[F, Msg]
+  )(using F: Async[F]): F[Nothing] = Dispatcher.sequential[F].use { dispatcher =>
+    (F.ref(ModelHolder(initModel, true)), AtomicCell[F].of(List.empty[(String, F[Unit])]), Channel.unbounded[F, Msg])
+      .flatMapN { (model, currentSubs, msgs) =>
```

Comment:
Nitpicking, but `msgs` sounds like the result rather than the conduit and forces a bit of mental gymnastics.

Later you're doing `msgs.send(_)` which reads like "Ok yes, send the messages... no wait... the message was in the foreach, this is sending those _into_ msgs...right?"

Reply:
Not nitpicking at all, you are right :) `msgChannel` maybe?